### ISSUE TITLE
Github Desktop and AppLocker-Rules

### ISSUE THREAD

**Issue by kabepst** [desktop/desktop#16604]
Problem in short: Github Desktop won't run when AppLocker-Enforcement Rules are in place. This seems to be due to the fact (perhaps git 3rd party binaries) that not all binaries shipped with Github Desktop are signed by Github itself. 

At my working place we had used Github Desktop for months without any problems (installed to %localappdata% via the Github Desktop MSI Installer for All-Users). Currently we are doing some QA-Tests for restricting Applications via Windows GPO-AppLocker Rules. What we tried:

1. Enable Default App-Locker Rules and update local policy on client (ignore Rules without "Default" in it's name) --> Screenshot 1 is from Executable Rules and Screenshot 2 from Packaged app Rules
![image](https://user-images.githubusercontent.com/70746076/233340227-bbc213a8-0cdb-4fa4-ac4d-59b49d7d8ca3.png)
![image](https://user-images.githubusercontent.com/70746076/233340417-d4294a10-d30a-4e5f-aedc-a53da76f54fa.png)

2. Try to run Github Desktop: Will be blocked by AppLocker as expected
3. Add the "Allow Github Desktop" Rule you already have seen above. This rule is a "Publisher based" Conditional Rule:
![image](https://user-images.githubusercontent.com/70746076/233341823-a11382ad-de56-4388-9deb-9a605c70c349.png)

4. Update Policy on client and try to run Github Desktop again. Github Desktop will start, but the mentioned error ("Failed to execute installGlobalLFSFilter: UNKNOWN") popus. Also I'm not able to clone/push/pull from any repository if i click on Close in the error message afterwards.(Repositories are located in C:\Github\ in this example, no OneDrive or similiar things in place).
 
For me this looks like there is another binary or simliar that Github Desktop tries to run which is signed differently than the Github Desktop.exe itself (see the rule screenshot above) and Windows fails to display the default "This App is blocked by your Administrator" message as it would happen when running the exe directly without an AppLocker Exception. 

What could be the root cause and fix it:

- Is there a git.exe or simliar binary which will be called when git commands are running within the %localappdata%\GithubDesktop Folder? If so: Can you please provide me the correct paths to them? This would allow me to create yet another rule and allow the execution of this binary. 
- Or perhaps: Is there a "recommended" AppLocker Configuration from Microsoft itself for Github Desktop ? For example see this for Microsoft Teams: https://learn.microsoft.com/en-us/microsoftteams/applocker-in-teams

If you need any log files or similiar things for further analyse don't hesitate to get in touch with me. I'm happy to help to find the root cause of this and fix it :)

_Originally posted by @kabepst in https://github.com/desktop/desktop/issues/10961#issuecomment-1516106619_

**Comment by kabepst** [comment]
So I've done another quick test and I'm quite sure it has to be the explained problem (running a binary which is signed by another Publisher as GithubDesktop.exe). For testing purposes I've created a path-exception (allow everything within %localappdata%\GithubDesktop) and there we go: Github Desktops is once again working as expected.

The rule itself:
![image](https://user-images.githubusercontent.com/70746076/234570867-0801a3c0-45b8-4ad4-bef4-4bce17bc42b0.png)

However:

- I would really like to avoid to have a "path-based" exception within the AppLocker-rules to a path which is "writeable" by the enduser (as people could use this folder to bypass AppLocker-Rules, for example: just copy a standalone binary to this folder and execute it..)
- I've done a quick search within the folders (*.exe) and there are plenty of them. So creating a rule for each of them isn't an option imho BUT: If there would be a list of all signed publishers (I hope there are only 2 or 3 ...) for those exe's it could be an option (because of that I referenced the MS Teams example above as it seems that Microsoft does indeed provided such lists for other products, perhaps this is also possible for Github Desktop?)
- Another option would be (at least for me) to move the whole installation to C:\Program Files or similiar folder. But this does not seem to be supported by Github Desktop if I'm right?

Originally posted by @kabepst in https://github.com/desktop/desktop/issues/10961#issuecomment-1516127308

**Comment by sergiou87** [comment]
That's interesting, AFAIK we were signing all included binaries with GitHub's certificate, but I've found at least two executables not signed by us:

![image](https://user-images.githubusercontent.com/1083228/234587195-29c59725-05a7-4c97-901c-84612933dcc8.png)

Could you try and add an exception for that specific folder and see if it's enough? I will investigate why they're not signed.

Thanks for your report and great catch!

**Comment by kabepst** [comment]
Hi Sergio,

As requested I've changed the exception from:
`C:\Users\*\AppData\Local\GitHubDesktop\*`

to:
`C:\Users\*\AppData\Local\GitHubDesktop\*\resources\app\git\cmd\*`

When starting Github I do get those errors now (which seem to be the same as without any path-exceptions but this time Github Desktop does indeed show a correct error message):

- ![github01_specificPath](https://user-images.githubusercontent.com/70746076/234596294-8612c00a-09a8-4b80-a2de-2ea278a33569.PNG)
  -  English: error launching git: This application was blocked by group policies. Contact your system-administrator to get additional infos.
- ![github02_specificPath](https://user-images.githubusercontent.com/70746076/234596316-857a8698-28d5-4f9f-8d3f-01e0b250a16e.PNG)

Looking into the logs for AppLocker (sry for that I could have done this already before opening this issue) I've found another occurance of that git.exe: ("Running of <path> was blocked"):
![image](https://user-images.githubusercontent.com/70746076/234597189-92b75f88-ea7c-4b31-9985-be4d8d697558.png)

So there we go with an  additional exception:

- `C:\Users\*\AppData\Local\GitHubDesktop\*\resources\app\git\mingw64\bin\*`
- When I try to open Github Desktop now, I do get the following error:
  - ![image](https://user-images.githubusercontent.com/70746076/234602038-4a680d4e-6ce2-454d-a2f4-23ce68d26f88.png)

But there are no more log-entries which indicate a blocked binary or similar thing.  The Rules in place are:

- Path-Exception for `C:\Users\*\AppData\Local\GitHubDesktop\*\resources\app\git\cmd\*`
- Path-Exception for `C:\Users\*\AppData\Local\GitHubDesktop\*\resources\app\git\mingw64\bin\*`
- Publisher-Exception:
  - ![image](https://user-images.githubusercontent.com/70746076/234604094-0f7e2186-5068-486f-91e0-7d5ef5208176.png)


However I was able to solve this issue by changing the above Publisher-Exception (the 2 Path-Exceptions are still in place and needed as expected) to (remove the Product name):
![image](https://user-images.githubusercontent.com/70746076/234604742-c1e43071-6ab8-499d-9884-75305ed8b0e5.png)

Please see my notes below why this worked. This is not related to the issue we're discussing here. So I'm wondering if I should open another issue for this. Also please not that I've only done a quick test by pull & push to/from an existing repo and cloning a new remote repo. Both worked without any issues now.

----

I've came across another thing which could be worth to have a look at when were already talking about signing binaries:

- The GitHubDesktop.exe is signed by the Publisher "GITHUB, INC" and with the product name "GITHUB DESKTOP"
- The Update.exe (same place) is signed by the Publisher "GITHUB, INC" and none Product name

As I'm not an expert in this field, this could be correct but if you follow the best-practices for creating AppLocker-Rules you will get (by opening an example exe) a Rule that looks like that:

![image](https://user-images.githubusercontent.com/70746076/234599756-84388cae-7f5b-4468-aad6-07ad685a3024.png)

As the Update.exe does not have a value for "Product name" it would still be blocked (I didn't ran into this before because of my global exception for the whole github-folder created during testing). Is there a reason this property has no value for the Updater.exe ? If not I would like to ask to include it, to avoid to create yet another rule (or excluding the Product name from the rule itself). But please note: This could also be a more "AppLocker" specific problem and not a "glitch" within the signing process for Github Desktop binaries. Also I'm happy to open another issue for this specific case if you prefer. Just let me know what you think about this :)

Edit: I went through this in my mind again and I think this is not an issue related to binary signing, but more an issue with properties/flags during compiling/build time perhaps? The "Product Name" is of course not part of the certificate itself (my fault). But the question remains: As I'm a sysadmin and not a developer, I can't tell if it is correct or not to leave out the Product name for Updater.exe while it is there for GitHubDesktop.exe

Cheers Pascal

**Comment by sergiou87** [comment]
Thank you for your detailed reply!

I've checked how Microsoft Teams is signed and I see the old version's `Update.exe` file does have a proper product name (which matches the other exes) but the new version of Teams has a different product name for the updater, hence why Microsoft, in that document you linked, adds 1 certificate but 2 products (Teams itself and its updater).

Right now we're handing off both setting that product name and signing all binaries to [electron-winstaller](https://github.com/electron/windows-installer). I will take a look when I have a chance and see if it's a known (maybe even solved!) bug or what.

**Comment by isd94techdir** [comment]
I don't have any details to add, just that I can't even get GitHub Desktop to install with AppLocker enabled. I tried to do the rule that kabepst has but I still can't get it to install. If there's any info I can add to help, let me know.

**Comment by kabepst** [comment]
@isd94techdir : You can find my shrinked (Default Rules & Github specific ones only) AppLocker Config attached. You can Import this into your GPO by doing a right-click on "AppLocker" and select "Import" within Group Policy Management Editor. But be aware that this will replace (delete all existing rules) your current rules. So make sure to not Import this into your actual productive  gpo. 

If that won't work, make sure the paths arre correct or adjust them if anything is different from your environment. Also please note: I wasn't able to test a fresh installation yet (as we already had Github Desktop in use when enabling AppLocker) so perhaps you ran into another problem. I will try to do a quick test within the next days and get back to you :)

And just for testing: Make sure you're executing the installer from a path/folder which isn't blocked by AppLocker (for example copy the installer file into %PROGRAMFILES% and execute it from there). I've seen some strange issues in the past (related to AppLocker in general and not Github Desktop specific) when trying to run a installer from a blocked-path. Never found out why to be honest and still have no clue what's going on there.

[appLocker_shrinkedGithubDesktopOnly.zip](https://github.com/desktop/desktop/files/11386471/appLocker_shrinkedGithubDesktopOnly.zip)

**Comment by kabepst** [comment]
> Thank you for your detailed reply!
> 
> I've checked how Microsoft Teams is signed and I see the old version's `Update.exe` file does have a proper product name (which matches the other exes) but the new version of Teams has a different product name for the updater, hence why Microsoft, in that document you linked, adds 1 certificate but 2 products (Teams itself and its updater).
> 
> Right now we're handing off both setting that product name and signing all binaries to [electron-winstaller](https://github.com/electron/windows-installer). I will take a look when I have a chance and see if it's a known (maybe even solved!) bug or what.

@sergiou87  Any news on that? :)

**Comment by sergiou87** [comment]
Sorry but no, we've been busy with accessibility improvements and didn't have time to look into this 😥

Right now we can't commit to any date for this 😭

**Comment by yawayl** [comment]
i wish i was a geek hacker new modern an true personator.  Road bus all the weighs over, in, and, out?

**Comment by kabepst** [comment]
> Sorry but no, we've been busy with accessibility improvements and didn't have time to look into this 😥
> 
> Right now we can't commit to any date for this 😭

Is there anything I can help with atm? IMHO this issue (at least the thing that not all binaries are signed by github itself) should be escalated/prioritized as Github Desktop is not usable at all (without investing a huge amount of time in testing/debuging) in managed-environments currently. AppLocker isn't that new and is widely adopted in the wild, at least for me :) In my opion Github should be interessed to support AppLocker "out of the box" (but that is more a in-general problem with electron and it's default installation paths) or at least have an FAQ/ReadMe Section on it's website which explains how to properly configure AppLocker-Exceptions for Github Desktop so not everyone has to go through the same struggle (If this is an option I'm happy to prepare a draft document for you as I could copy/paste it mostly from my own  internal documentation).

**Comment by sergiou87** [comment]
@kabepst I guess the problem is we haven't received many reports related to AppLocker 😞

However I'm doing a quick search and it'd be useful to have all issues with references to it linked here:
- https://github.com/desktop/desktop/issues/17595
- https://github.com/desktop/desktop/issues/10961
- https://github.com/desktop/desktop/issues/4371
- https://github.com/desktop/desktop/issues/11329
- https://github.com/desktop/desktop/issues/4345
- https://github.com/desktop/desktop/issues/1086

Looking at that last issue I just linked, looks like VSCode switched from Squirrel to Inno Setup to provide machine-wide installers. If that's what's necessary, it will be hard for a small team like us to find time to work on it.

If we only need all binaries to be signed with GitHub's certificate, I think we just need to remove the existing signature (if that's possible) of the git binaries we bundle with the app before Desktop is packaged, and then Squirrel will sign all of them with GitHub's certificate.

If you want to try that out and/or write some documentation in our [known-issues file](https://github.com/desktop/desktop/blob/development/docs/known-issues.md), you will be more than welcome! 😄

**Comment by kabepst** [comment]
@sergiou87 Just wanted to let you know that I appreciate your effort and I've saved this thread to my todo-list to look at it within my next PTO since this error came up up first when doing stuff for my employer and as described it works there now. In short: Employer won't approve anymore time investing on this.

I will get back to this at the end of december and hopefully at least provide a resume what imho would be the "perfect" way to go in future to improve the overall behavior (more a theoretical approach then really possible to realize as you already laid out) and also prepare a section for the known-issues file how AppLocker needs to be configured to b able to run Github Desktop without any issues.

**Comment by rrjat4840** [comment]
jackpotjitnekatarika #jackpotkaisejite

**Comment by rrjat4840** [comment]
jackpotjitnekatarika #jackpotkaisejite

**Comment by rrjat4840** [comment]
jackpotjitnekatarika #jackpotkaisejite[](url)

**Comment by isikyus** [comment]
> \@kabepst I guess the problem is we haven't received many reports related to AppLocker 😞

Count me as another person seeing this kind of issue.

My employer is using Software Restriction Policies, which I believe are not quite the same as AppLocker, but we're having basically the same problem: GitHub Desktop installs a multitude of individual executables into a folder under AppData, and there isn't any easy way to allow all of them.

It looks like we will be able to get it working with a long list of path-based rules, but it's very time-consuming to set up, and we'll need to update it again for every new GitHub Desktop version. It really isn't a smooth process.

See also my comment https://github.com/desktop/desktop/issues/17127#issuecomment-2448702161
